### PR TITLE
feat: preferNonPending

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds new feature `preferNonPending `. If set to `true`, `stabilityDays` evaluation will be done as part of lookup filtering, so that non-satisfying releases are not proposed when satisfying releases are available.

## Context:

#7603

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Example log:

```
             "deps": [
               {
                 "depType": "dependencies",
                 "depName": "aws-sdk",
                 "currentValue": "2.838.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "depIndex": 0,
                 "updates": [
                   {
                     "fromVersion": "2.838.0",
                     "toVersion": "2.841.0",
                     "newValue": "2.841.0",
                     "bucket": "non-major",
                     "newMajor": 2,
                     "newMinor": 841,
                     "updateType": "minor",
                     "isSingleVersion": true,
                     "releaseTimestamp": "2021-02-09T19:19:49.824Z",
                     "skippedOverVersions": ["2.839.0", "2.840.0"],
                     "pendingVersions": ["2.842.0", "2.843.0"]
                   }
                 ],
                 "warnings": [],
                 "fixedVersion": "2.838.0",
                 "sourceUrl": "https://github.com/aws/aws-sdk-js"
               }
             ],
```

Note how `2.843.0` is latest but `2.841.0` is proposed instead, as stabilityDays is set to 3

File: lib/workers/repository/process/lookup/index.ts [hunk 1]
```diff
@@ -329,7 +330,47 @@ export async function lookupUpdates(
       const sortedReleases = bucketReleases.sort((r1, r2) =>
         versioning.sortVersions(r1.version, r2.version)
       );
-      const bucketRelease = sortedReleases.pop();
+      let bucketRelease: Release;
+      let skippedOverVersions: string[];
+      let pendingVersions: string[];
+      // TODO: add this to config definitions
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.ts:336]: 😏

File: lib/workers/repository/process/lookup/index.ts [hunk 2]
```diff
@@ -329,7 +330,47 @@ export async function lookupUpdates(
...
+            if (release.releaseTimestamp) {
+              const daysElapsed = Math.floor(
+                (new Date().getTime() -
+                  new Date(release.releaseTimestamp).getTime()) /
+                  ONE_DAY
+              );
+              if (daysElapsed < config.stabilityDays) {
+                pendingChecks.push(release.version);
+              }
+            }
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.ts:354]: This should be done with `luxon` as it's much easier and cleaner. can send a sample later

**Review by viceice — COMMENTED** [review]
Will do another more complete review later

File: lib/workers/repository/process/lookup/index.ts (hunk 2 above)
Comment (viceice) [reply lib/workers/repository/process/lookup/index.ts:354]: https://github.com/renovatebot/renovate/blob/6fa37294cfb103f5b7a3c8e8b186303bfa4842ce/lib/workers/branch/index.ts#L2
https://github.com/renovatebot/renovate/blob/6fa37294cfb103f5b7a3c8e8b186303bfa4842ce/lib/workers/branch/index.ts#L456-L457

File: lib/workers/repository/process/lookup/index.ts [hunk 3]
```diff
@@ -143,6 +146,24 @@ function getBucket(
...
+  config: LookupUpdateConfig,
+  release: Release
+): boolean {
+  if (!config.stabilityDays) {
+    return false;
+  }
+  if (!release?.releaseTimestamp) {
+    return false;
+  }
+  const releaseTimestamp = DateTime.fromISO(release.releaseTimestamp);
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.ts:159]: Needs a check for `isValid`. How should we handle invalid date strings? treat them as undefined?
maybe log a debug message and ignore?

https://moment.github.io/luxon/docs/class/src/datetime.js~DateTime.html#instance-get-isValid

Comment (rarkins) [reply lib/workers/repository/process/lookup/index.ts:159]: We should ideally validate them in the datasource index and discard if invalid. BTW I was wondering if `DateTime.fromISO()` is too restricted, e.g. compared to `new Date()`?

Comment (viceice) [reply lib/workers/repository/process/lookup/index.ts:159]: no, i think it's great, as all other code depoends on this too. So we should test and fix it in datasource as you suggested.

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
Blocked by #8695

**Comment by rarkins** [comment]
Moved to draft mode as I realised that `stabilityDays` will often be configured per-updateType and evaluating it in the filter would be too early.

**Comment by ibratoev** [comment]
@rarkins Maybe we can introduce another option for that. For example, `ignoreDays` (or `minimumAgeDays`) and allow that to be configurable only on the top level for the moment. This would allow for both options to be used. 

In my scenario, I just want to ignore all releases for the first 3-5 days to add some protection against potential malicious change that would be caught in a matter of days.

**Comment by rarkins** [comment]
Superseded by #9796

**Pull request closed without merging**